# 1.Ingest the data 

In [2]:
import pandas as pd 

 Reading the file using the Absolute path

In [10]:
data=pd.read_csv(r"D:\iNeuron_pvt_ltd\Machine_Learning\End_to_End_Project\notebooks\data\gemstone.csv")
data

id  carat        cut color clarity  depth  table     x     y  \
0            0   1.52    Premium     F     VS2   62.2   58.0  7.27  7.33   
1            1   2.03  Very Good     J     SI2   62.0   58.0  8.06  8.12   
2            2   0.70      Ideal     G     VS1   61.2   57.0  5.69  5.73   
3            3   0.32      Ideal     G     VS1   61.6   56.0  4.38  4.41   
4            4   1.70    Premium     G     VS2   62.6   59.0  7.65  7.61   
...        ...    ...        ...   ...     ...    ...    ...   ...   ...   
193568  193568   0.31      Ideal     D    VVS2   61.1   56.0  4.35  4.39   
193569  193569   0.70    Premium     G    VVS2   60.3   58.0  5.75  5.77   
193570  193570   0.73  Very Good     F     SI1   63.1   57.0  5.72  5.75   
193571  193571   0.34  Very Good     D     SI1   62.9   55.0  4.45  4.49   
193572  193572   0.71       Good     E     SI2   60.8   64.0  5.73  5.71   

           z  price  
0       4.55  13619  
1       5.05  13387  
2       3.50   2772  
3       2.71    666  
4       4.77  14453  
...      ...    ...  
193568  2.67   1130  
193569  3.47   2874  
193570  3.62   3036  
193571  2.81    681  
193572  3.48   2258  

[193573 rows x 11 columns]

In [ ]:
#Relative path
notebooks\data\gemstone.csv

In [ ]:
#Absolute path---> complete __path__
D:\iNeuron_pvt_ltd\Machine_Learning\End_to_End_Project\notebooks\data\gemstone.csv

## Introduction About the Data :

Dataset Source Link : https://www.kaggle.com/competitions/playground-series-s3e8/data?select=train.csv

The dataset The goal is to predict price of given diamond (Regression Analysis).

There are 10 independent variables (including id):


1. **id** : unique identifier of each diamond 

2. A **carat** is a unit of measurement for the weight of a diamond. One carat is equal to 0.200 grams or 1/5 gram. A carat is subdivided into 100 points, which allows for very precise measurements to the hundredth decimal place.

3. The GIA (Gemological Institute of America) has five **cut** grades for diamonds:
      * Excellent (EX): The best grade ---> here Premium
      *  Very Good (VG):
      *  Good (G):
      *  Fair (F):
      * Poor (P): The worst grade

4. GIA **COLOR** SCALE 
     1. colorless
        * D , E , F 
     2. Near colorless 
        * G ,H , I , J
     3. Faint 
        * K ,L ,M 
     4. Very Light 
        * N , O , P , Q , R
     5. Light     
        *  S , T , U , V , W , X, Y , Z

5. The GIA Diamond **Clarity** Scale has 6 categories, some of which are divided, for a total of 11 specific grades.

     1. **Flawless (FL)** No inclusions and no blemishes visible under 10x magnification

     2. **Internally Flawless (IF)** No inclusions visible under 10x magnification

     3. **Very, Very Slightly Included (VVS1 and VVS2)** Inclusions so slight they are difficult for a skilled grader to see under 10x magnification

     4. **Very Slightly Included (VS1 and VS2)** Inclusions are observed with effort under 10x magnification, but can be characterized as minor

     5. **Slightly Included (SI1 and SI2)** Inclusions are noticeable under 10x magnification

     6. **Included (I1, I2, and I3)** Inclusions are obvious under 10x magnification which may affect transparency and brilliance   

6. The **depth** of a diamond refers to its measurement from top to bottom, from the table on the top of the diamond to the culet at its base. The depth of any diamond is expressed as a percentage.


     * For a round diamond, an ideal depth percentage is between 59 and 62.6 percent
     * For a princess cut diamond, choose a diamond with an ideal depth of 68 to 74 percent.
     * For a cushion cut diamond, look for an ideal depth that’s between 61 and 68 percent.
     * For an Asscher or Emerald cut diamond, an ideal depth is between 61 to 68 percent.
     * For an oval cut diamond, an ideal depth is less than 68 percent.
     * For a pear shape diamond, an ideal diamond’s depth is less than 68 percent.
     * For a radiant cut diamond, an ideal depth is less than 67 percent.
     * For a heart shape diamond, make sure to choose a diamond with an ideal depth of 56 to 62 percent.
     * For a marquise diamond, an ideal depth range is between 58 and 62 percent of the total width of the diamond, while a very good cut will have a depth range of 56 to 57.9 or 62.1 to 66 percent.
   
        * Diamond Shape--------------Ideal Depth Percentage
        * Round Brilliant Cut--------59 to 62.6%
        * Princess Cut---------------68 to 74%
        * Cushion Cut----------------61 to 68%
        * Emerald Cut----------------61 to 68%
        * Asscher Cut----------------61 to 68%
        * Oval Cut-------------------< 68%
        * Pear Shape-----------------< 68%
        * Radiant Cut----------------< 67%
        * Heart Shape----------------56 to 62%
        * Marquise Cut---------------58 to 62%



7. A diamond’s table is the flat facet on its surface — the large, flat surface facet that you can see when you look at the diamond from above. As the largest facet on a diamond, the table plays a major role in determining how brilliant (sparkly) the diamond is. 

        * Diamond Shape---------------Ideal Table Percentage	
        * Round Brilliant---------------Cut	54 to 57%	
        * Princess Cut---------------69 to 75%	
        * Cushion Cut---------------< 68%
        * Emerald Cut---------------60 to 68%	
        * Asscher Cut---------------60 to 68%	
        * Oval Cut---------------53 to 63%	
        * Pear Shape---------------53 to 65%	
        * Radiant Cut---------------61 to 69%	
        * Heart Shape---------------56 to 62%	
        * Marquise Cut---------------53 to 63%
8.  X : Length of the Diamond in mm.
9.  Y : Width of the Diamond in mm
10. Z : Height of the Diamond in mm.
11. Price of Diamond





## Exploratory Data Analysis
--------------------------------

### 1. size of the data


In [18]:
data.shape

(193573, 11)

### 2. how the data looks likes

In [19]:
data.head(5)

id  carat        cut color clarity  depth  table     x     y     z  price
0   0   1.52    Premium     F     VS2   62.2   58.0  7.27  7.33  4.55  13619
1   1   2.03  Very Good     J     SI2   62.0   58.0  8.06  8.12  5.05  13387
2   2   0.70      Ideal     G     VS1   61.2   57.0  5.69  5.73  3.50   2772
3   3   0.32      Ideal     G     VS1   61.6   56.0  4.38  4.41  2.71    666
4   4   1.70    Premium     G     VS2   62.6   59.0  7.65  7.61  4.77  14453

In [20]:
data.tail(5)

id  carat        cut color clarity  depth  table     x     y  \
193568  193568   0.31      Ideal     D    VVS2   61.1   56.0  4.35  4.39   
193569  193569   0.70    Premium     G    VVS2   60.3   58.0  5.75  5.77   
193570  193570   0.73  Very Good     F     SI1   63.1   57.0  5.72  5.75   
193571  193571   0.34  Very Good     D     SI1   62.9   55.0  4.45  4.49   
193572  193572   0.71       Good     E     SI2   60.8   64.0  5.73  5.71   

           z  price  
193568  2.67   1130  
193569  3.47   2874  
193570  3.62   3036  
193571  2.81    681  
193572  3.48   2258

### 3. what are the dtype of col

In [21]:
data.dtypes

id           int64
carat      float64
cut         object
color       object
clarity     object
depth      float64
table      float64
x          float64
y          float64
z          float64
price        int64
dtype: object

### 4. how data looks like mathematically

In [26]:
data.describe()

id          carat          depth          table  \
count  193573.000000  193573.000000  193573.000000  193573.000000   
mean    96786.000000       0.790688      61.820574      57.227675   
std     55879.856166       0.462688       1.081704       1.918844   
min         0.000000       0.200000      52.100000      49.000000   
25%     48393.000000       0.400000      61.300000      56.000000   
50%     96786.000000       0.700000      61.900000      57.000000   
75%    145179.000000       1.030000      62.400000      58.000000   
max    193572.000000       3.500000      71.600000      79.000000   

                   x              y              z          price  
count  193573.000000  193573.000000  193573.000000  193573.000000  
mean        5.715312       5.720094       3.534246    3969.155414  
std         1.109422       1.102333       0.688922    4034.374138  
min         0.000000       0.000000       0.000000     326.000000  
25%         4.700000       4.710000       2.900000     951.000000  
50%         5.700000       5.720000       3.530000    2401.000000  
75%         6.510000       6.510000       4.030000    5408.000000  
max         9.650000      10.010000      31.300000   18818.000000

In [22]:
data.describe().T

count          mean           std    min       25%       50%  \
id     193573.0  96786.000000  55879.856166    0.0  48393.00  96786.00   
carat  193573.0      0.790688      0.462688    0.2      0.40      0.70   
depth  193573.0     61.820574      1.081704   52.1     61.30     61.90   
table  193573.0     57.227675      1.918844   49.0     56.00     57.00   
x      193573.0      5.715312      1.109422    0.0      4.70      5.70   
y      193573.0      5.720094      1.102333    0.0      4.71      5.72   
z      193573.0      3.534246      0.688922    0.0      2.90      3.53   
price  193573.0   3969.155414   4034.374138  326.0    951.00   2401.00   

             75%        max  
id     145179.00  193572.00  
carat       1.03       3.50  
depth      62.40      71.60  
table      58.00      79.00  
x           6.51       9.65  
y           6.51      10.01  
z           4.03      31.30  
price    5408.00   18818.00

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193573 entries, 0 to 193572
Data columns (total 11 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       193573 non-null  int64  
 1   carat    193573 non-null  float64
 2   cut      193573 non-null  object 
 3   color    193573 non-null  object 
 4   clarity  193573 non-null  object 
 5   depth    193573 non-null  float64
 6   table    193573 non-null  float64
 7   x        193573 non-null  float64
 8   y        193573 non-null  float64
 9   z        193573 non-null  float64
 10  price    193573 non-null  int64  
dtypes: float64(6), int64(2), object(3)
memory usage: 16.2+ MB


### 5.Checking whether there is missing values

In [13]:
data.isnull().sum()

id         0
carat      0
cut        0
color      0
clarity    0
depth      0
table      0
x          0
y          0
z          0
price      0
dtype: int64

### 6.Checking for duplicated values

In [28]:
data.duplicated().sum()

0

In [29]:
data[data.duplicated()]

Empty DataFrame
Columns: [id, carat, cut, color, clarity, depth, table, x, y, z, price]
Index: []

### 7.Checking for unique values

In [32]:
data.nunique()

id         193573
carat         248
cut             5
color           7
clarity         8
depth         153
table         108
x             522
y             521
z             349
price        8738
dtype: int64

### 8.Dropping the id coloumn

Since it is not contributing to model building or prediction

In [34]:
new_data=data.drop(labels={'id'},axis=1)
new_data

carat        cut color clarity  depth  table     x     y     z  price
0        1.52    Premium     F     VS2   62.2   58.0  7.27  7.33  4.55  13619
1        2.03  Very Good     J     SI2   62.0   58.0  8.06  8.12  5.05  13387
2        0.70      Ideal     G     VS1   61.2   57.0  5.69  5.73  3.50   2772
3        0.32      Ideal     G     VS1   61.6   56.0  4.38  4.41  2.71    666
4        1.70    Premium     G     VS2   62.6   59.0  7.65  7.61  4.77  14453
...       ...        ...   ...     ...    ...    ...   ...   ...   ...    ...
193568   0.31      Ideal     D    VVS2   61.1   56.0  4.35  4.39  2.67   1130
193569   0.70    Premium     G    VVS2   60.3   58.0  5.75  5.77  3.47   2874
193570   0.73  Very Good     F     SI1   63.1   57.0  5.72  5.75  3.62   3036
193571   0.34  Very Good     D     SI1   62.9   55.0  4.45  4.49  2.81    681
193572   0.71       Good     E     SI2   60.8   64.0  5.73  5.71  3.48   2258

[193573 rows x 10 columns]

### 9.Segregating the data into categorical and numerical 

In [37]:
cat=data.columns[data.dtypes == 'object']
cat

Index(['cut', 'color', 'clarity'], dtype='object')

In [38]:
num=data.columns[data.dtypes != 'object']
num

Index(['id', 'carat', 'depth', 'table', 'x', 'y', 'z', 'price'], dtype='object')

In [43]:
data[cat]

cut color clarity
0         Premium     F     VS2
1       Very Good     J     SI2
2           Ideal     G     VS1
3           Ideal     G     VS1
4         Premium     G     VS2
...           ...   ...     ...
193568      Ideal     D    VVS2
193569    Premium     G    VVS2
193570  Very Good     F     SI1
193571  Very Good     D     SI1
193572       Good     E     SI2

[193573 rows x 3 columns]

In [44]:
data[cat].describe().T

count unique    top   freq
cut      193573      5  Ideal  92454
color    193573      7      G  44391
clarity  193573      8    SI1  53272

In [40]:
data[num]

id  carat  depth  table     x     y     z  price
0            0   1.52   62.2   58.0  7.27  7.33  4.55  13619
1            1   2.03   62.0   58.0  8.06  8.12  5.05  13387
2            2   0.70   61.2   57.0  5.69  5.73  3.50   2772
3            3   0.32   61.6   56.0  4.38  4.41  2.71    666
4            4   1.70   62.6   59.0  7.65  7.61  4.77  14453
...        ...    ...    ...    ...   ...   ...   ...    ...
193568  193568   0.31   61.1   56.0  4.35  4.39  2.67   1130
193569  193569   0.70   60.3   58.0  5.75  5.77  3.47   2874
193570  193570   0.73   63.1   57.0  5.72  5.75  3.62   3036
193571  193571   0.34   62.9   55.0  4.45  4.49  2.81    681
193572  193572   0.71   60.8   64.0  5.73  5.71  3.48   2258

[193573 rows x 8 columns]

In [46]:
data[num].describe().T

count          mean           std    min       25%       50%  \
id     193573.0  96786.000000  55879.856166    0.0  48393.00  96786.00   
carat  193573.0      0.790688      0.462688    0.2      0.40      0.70   
depth  193573.0     61.820574      1.081704   52.1     61.30     61.90   
table  193573.0     57.227675      1.918844   49.0     56.00     57.00   
x      193573.0      5.715312      1.109422    0.0      4.70      5.70   
y      193573.0      5.720094      1.102333    0.0      4.71      5.72   
z      193573.0      3.534246      0.688922    0.0      2.90      3.53   
price  193573.0   3969.155414   4034.374138  326.0    951.00   2401.00   

             75%        max  
id     145179.00  193572.00  
carat       1.03       3.50  
depth      62.40      71.60  
table      58.00      79.00  
x           6.51       9.65  
y           6.51      10.01  
z           4.03      31.30  
price    5408.00   18818.00

In [48]:
data[cat]['clarity'].value_counts()

clarity
SI1     53272
VS2     48027
VS1     30669
SI2     30484
VVS2    15762
VVS1    10628
IF       4219
I1        512
Name: count, dtype: int64

In [49]:
data[cat]['cut'].value_counts()

cut
Ideal        92454
Premium      49910
Very Good    37566
Good         11622
Fair          2021
Name: count, dtype: int64

In [50]:
data[cat]['color'].value_counts()

color
G    44391
E    35869
F    34258
H    30799
D    24286
I    17514
J     6456
Name: count, dtype: int64